zun/container/k8s/driver.py

```diff
@@ -60,6 +66,48 @@ def _pod_ips(pod):
         return []
     return [p.ip for p in pod.status.pod_i_ps]
 
+class WSFileManager:
+    """
+    WS wrapper to manage read and write bytes in K8s WSClient
+    source: https://stackoverflow.com/a/64374553
```

Comment:
kubernetes exposes a "pod exec" api, but not a "file upload/download" one.
https://github.com/kubernetes-client/python/blob/master/examples/pod_exec.py
https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md#connect_get_namespaced_pod_exec

File upload/download can be implemented by piping file content through this exec API, to a command that writes it back to a file.

Instead of a synchronous `get` call, kubernetes APIs can also be accessed via a streaming, websocket-based method.
https://github.com/kubernetes-client/python/blob/master/kubernetes/base/stream/stream.py#L20-L22

This code uses those two abstractions to open a websocket connection, and attempt to read data from stdout and stderr until the connection is closed, by means of repeatedly calling `read_bytes`

Reply:
I added a longer commit description that helps to understand the purpose of WSFileManager